### PR TITLE
Code optimizations to reduce disk writes on SONiC switches

### PR DESCRIPTION
**Pull request by assrinivasan**
#### Why I did it
This PR contains various optimizations to reduce disk writes on SONiC switches as mentioned in the [High-level Design](https://github.com/sonic-net/SONiC/pull/1838). 

This contributes to the overall reduction in data written to the disk over time, which can help extend the life of the disk and reduce the risk of disk failure.

##### Work item tracking
- Microsoft ADO **(number only)**: 28300356

#### How I did it
1. Moved supervisord logs within swss to tmpfs
2. Optimized away repeated disk writes by vtysh in the bgp container
3. Moved monit and logrotate writers to tmpfs

#### How to verify it

Flash image with these optimizations onto a sonic switch and check the following:

1. Within swss container, supervisord logs are written to /dev/shm/supervisor/
2. Within bgp container, verify that ~/.history_frr file is not written into repeatedly with `show bgp summary json`
3. Verify that monit state file is in /dev/shm/monit
4. Verify that logrotate status file is in /dev/shm/logrotate

#### Which release branch to backport (provide reason below if selected)

- [x] 202411

#### Tested branch (Please provide the tested image version)

- [x] 20201231.119
- [x] 20220531.47
- [x] 20230531.19

#### Description for the changelog

Code optimizations to reduce disk writes on SONiC switches

### CODE REVIEW

File: files/build_templates/sonic_debian_extension.j2 [hunk 1]
```diff
@@ -104,6 +104,10 @@ sudo dpkg --root=$FILESYSTEM_ROOT -i $debs_path/sonic-nettools_*.deb || \
     sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get -y install -f
 sudo setcap 'cap_net_raw=+ep' $FILESYSTEM_ROOT/usr/bin/wol
 
+# This is needed for moving monit logs, state and logrotate status to tmpfs
+sudo mkdir -p $FILESYSTEM_ROOT/dev/shm/monit/
+sudo mkdir -p $FILESYSTEM_ROOT/dev/shm/logrotate/
```
Comment (saiarcot895) [inline files/build_templates/sonic_debian_extension.j2:109]: This isn't going to be present on bootup. At bootup, a tmpfs filesystem is mounted at `/dev/shm`.

**Comment by assrinivasan** [comment]
@keboliu @Junchao-Mellanox @dgsudharsan, please help review this PR, thank you in advance.

File: dockers/docker-base-bullseye/etc/supervisor/supervisord.conf
Comment (Junchao-Mellanox) [inline dockers/docker-base-bullseye/etc/supervisor/supervisord.conf:1]: Only bullseye? How about other docker images?

File: files/build_templates/sonic_debian_extension.j2 (hunk 1 above)
Comment (assrinivasan) [reply files/build_templates/sonic_debian_extension.j2]: Changed in latest per our discussion.

File: dockers/docker-base-bullseye/etc/supervisor/supervisord.conf
Comment (assrinivasan) [reply dockers/docker-base-bullseye/etc/supervisor/supervisord.conf:1]: Thank you for catching the miss. Added to bookworm.

File: platform/components/docker-gbsyncd-broncos.mk [hunk 2]
```diff
@@ -11,6 +11,7 @@ DOCKER_GBSYNCD_BRONCOS = $(DOCKER_GBSYNCD_BRONCOS_STEM).gz
 DOCKER_GBSYNCD_BRONCOS_DBG = $(DOCKER_GBSYNCD_BRONCOS_STEM)-$(DBG_IMAGE_MARK).gz
 
 $(DOCKER_GBSYNCD_BRONCOS)_FILES += $(SUPERVISOR_PROC_EXIT_LISTENER_SCRIPT)
+$(DOCKER_GBSYNCD_BRONCOS)_FILES += $(DSSERVE)
```
Comment (prgeor) [inline platform/components/docker-gbsyncd-broncos.mk:14]: @assrinivasan how is this change relevant to the PR topic?

**Review by prgeor — DISMISSED** [review]

File: dockers/docker-orchagent/docker-init.j2 [hunk 3]
```diff
@@ -3,7 +3,7 @@
 mkdir -p /etc/swss/config.d/
 mkdir -p /etc/supervisor/
 mkdir -p /etc/supervisor/conf.d/
-
+mkdir -p /dev/shm/supervisor/
```
Comment (saiarcot895) [inline dockers/docker-orchagent/docker-init.j2:6]: Is orchagent the only container that needs this? Or is it handled for the other containers by some other means?

Comment (assrinivasan) [reply dockers/docker-orchagent/docker-init.j2:6]: Orchagent was the only container we found was contributing towards disk writes. I'm hesitant to "fix it if it aint broke", so to speak.

Comment (assrinivasan) [reply dockers/docker-orchagent/docker-init.j2:6]: made /var/log/supervisor directory tmpfs on all containers after internal discussion

File: files/image_config/logrotate/logrotateOverride.conf [hunk 4]
```diff
@@ -1,2 +1,6 @@
 [Unit]
-Requires=logrotate-config.service
\ No newline at end of file
+Requires=logrotate-config.service
+
+[Service]
+ExecStart=
```
Comment (prgeor) [inline files/image_config/logrotate/logrotateOverride.conf:5]: @assrinivasan remove?

Comment (assrinivasan) [reply files/image_config/logrotate/logrotateOverride.conf:5]: No @prgeor. We need this to have logrotate write its status to /dev/shm/logrotate.

File: files/initramfs-tools/union-mount.j2 [hunk 5]
```diff
@@ -212,6 +212,9 @@ mkdir -p ${rootmnt}/boot
 mkdir -p ${rootmnt}/host/$image_dir/boot
 mount --bind ${rootmnt}/host/$image_dir/boot ${rootmnt}/boot
 
+## Mount the /tmp directory as tmpfs
```
[the remark this answers is not in the document]
Comment (assrinivasan) [reply files/initramfs-tools/union-mount.j2:215]: We are currently allocating 25% of total RAM towards /tmp directory on all platforms.

File: files/image_config/logrotate/logrotateOverride.conf (hunk 4 above)
Comment (saiarcot895) [reply files/image_config/logrotate/logrotateOverride.conf:5]: The `ExecStart=` line is intentional. It's to tell systemd to clear out the previous value set for `ExecStart`. Without this line, systemd will append the new command, and the service will have two commands to run; in this case, logrotate would then run twice.

**Review by saiarcot895 — APPROVED** [review]

**Review by prgeor — APPROVED** [review]

**Comment by kperumalbfn** [comment]
@assrinivasan Could you create a separate PR for 202411 branch?

**Comment by kperumalbfn** [comment]
@prgeor @saiarcot895 Could you check and confirm if this is needed for 202411. If so, please create a separate PR to resolve the conflicts and we can push to 202411.
